File: backend/src/domain/models/export.py
import enum
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.config.database import Base


class ExportFormat(str, enum.Enum):
    CSV = "csv"
    PDF = "pdf"
    EXCEL = "excel"
    JSON = "json"


class ExportStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class ExportType(str, enum.Enum):
    TRANSACTIONS = "transactions"
    BUDGET_REPORT = "budget_report"
    ANNUAL_REPORT = "annual_report"
    MONTHLY_REPORT = "monthly_report"
    CUSTOM = "custom"


class Export(Base):
    __tablename__ = "exports"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    type: Mapped[ExportType] = mapped_column(Enum(ExportType), nullable=False)
    format: Mapped[ExportFormat] = mapped_column(Enum(ExportFormat), nullable=False)
    status: Mapped[ExportStatus] = mapped_column(
        Enum(ExportStatus), default=ExportStatus.PENDING, nullable=False
    )
    file_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    file_size_bytes: Mapped[int | None] = mapped_column(nullable=True)
    filters: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    error_message: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    user: Mapped["User"] = relationship(back_populates="exports")

File: lib/features/reports/domain/entities/export.dart
enum ExportFormat { csv, pdf, excel, json }
enum ExportStatus { pending, processing, completed, failed }
enum ExportType { transactions, budgetReport, annualReport, monthlyReport, custom }

class Export {
  final int id;
  final int userId;
  final ExportType type;
  final ExportFormat format;
  final ExportStatus status;
  final String? fileUrl;
  final int? fileSizeBytes;
  final String? errorMessage;
  final DateTime createdAt;
  final DateTime? completedAt;

  const Export({
    required this.id,
    required this.userId,
    required this.type,
    required this.format,
    this.status = ExportStatus.pending,
    this.fileUrl,
    this.fileSizeBytes,
    this.errorMessage,
    required this.createdAt,
    this.completedAt,
  });
}
